src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+            case MPIR_COLL_GENERIC_KIND_SEND:
+            case MPIR_COLL_GENERIC_KIND_RECV:
+                if (mpid_req0 && MPIR_Request_is_complete(mpid_req0)) {
+                    MPIR_Request_free(mpid_req0);
+                    vtxp->mpid_req[0] = NULL;
+#ifdef MPIR_COLL_DEBUG
+                    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"  --> GENERIC transport (kind=%d) complete\n", vtxp->kind));
+                    if (vtxp->nbargs.sendrecv.count >= 1)
+                        MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"data sent/recvd: %d\n", *(int *) (vtxp->nbargs.sendrecv.buf)));
+#endif
+                    /* record vertex completion */
+                    MPIR_COLL_GENERIC_record_vtx_completion(vtxp, sched);
+                } else
+                    MPIR_COLL_GENERIC_record_vtx_issue(vtxp, sched);   /* record it again as issued */
+                break;
+            case MPIR_COLL_GENERIC_KIND_MULTICAST:
+                for (i=vtxp->nbargs.multicast.last_complete+1; i<vtxp->nbargs.multicast.num_destinations; i++) {
+                    if(MPIR_Request_is_complete(vtxp->nbargs.multicast.mpir_req[i])){
+                        MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"  --> GENERIC transport multicast task %d complete\n", i));
+                        MPIR_Request_free(vtxp->nbargs.multicast.mpir_req[i]);
+                        vtxp->nbargs.multicast.mpir_req[i] = NULL;
+                        vtxp->nbargs.multicast.last_complete = i;
+                    }
+                    else /* we are only checking in sequence, hence break out at the first incomplete send */
+                        break;
+                }
+                /* if all the sends have completed, mark this vertex as complete */
+                if (i==vtxp->nbargs.multicast.num_destinations)
+                    MPIR_COLL_GENERIC_record_vtx_completion(vtxp, sched);
+                else
+                    MPIR_COLL_GENERIC_record_vtx_issue(vtxp, sched);
+                break;
+            case MPIR_COLL_GENERIC_KIND_RECV_REDUCE:
+                /* check for recv completion */
+                if (mpid_req0 && MPIR_Request_is_complete(mpid_req0)) {
+                    MPIR_Request_free(mpid_req0);
+                    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"recv in recv_reduce completed\n"));
+                    vtxp->mpid_req[0] = NULL;
+                }
+
+                /* check for reduce completion */
+                if (mpid_req1 && MPIR_Request_is_complete(mpid_req1)) {
+                    MPIR_Request_free(mpid_req1);
+                    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"reduce in recv_reduce also completed\n"));
+                    vtxp->mpid_req[1] = NULL;
+
+                    /* recv_reduce is now complete because reduce can complete
+                     * only if the corresponding recv is complete */
+#ifdef MPIR_COLL_DEBUG
```

Comment:
If you're going to use the `MPIR_COLL_DEBUG` macros here, put them everywhere. However, since you're already using the `MPL_` debug macros, I think this is unnecessary. I think the compiler can figure out that the code block below is a noop if debug output is turned off.

Reply:
But what if there is actual code that is executed before the MPL_Print is called. I don't want to execute that code unless debugging is turned on. What is the MPL_ equivalent of MPIR_COLL_DEBUG?

Reply:
I'm not sure it's really a problem. Doesn't the `MPL_DBG_MSG_FMT` macro become a no-op of debug statements are turned off? If that's the case, then surely the compiler would remove the if statement.

Reply:
In this case, it is obvious. What is there is code where a string is constructed by running some for loops, and then that string is printed. Is it safe to assume that compiler will remove those for loops also when print is no-op.

Reply:
Yeah, that's obviously true. I agree that there are cases where it's necessary. I guess I should retract what I said about using them everywhere and just say that I don't think they're necessary here.

Reply:
I think we should use the MPL equivalent in any case.